python/lsst/fgcmcal/fgcmOutputIlluminationCorrection.py

```diff
@@ -0,0 +1,348 @@
+import dataclasses
+
+from lsst.afw.image import ExposureF, FilterLabel
+from lsst.daf.butler import Timespan
+from lsst.pipe.base import PipelineTaskConnections, PipelineTaskConfig, connectionTypes, PipelineTask, Struct
+import lsst.pex.config
+
+from .fgcmOutputProducts import FgcmOutputProductsTask
+from .utilities import computePixelAreaFieldDetector, computeReferencePixelScale
+
+__all__ = ["FgcmOutputIlluminationCorrectionConfig", "FgcmOutputIlluminationCorrectionTask"]
+
+
+class FgcmOutputIlluminationCorrectionConnections(
+    PipelineTaskConnections,
+    dimensions=("instrument", "detector"),
+    defaultTemplates={"cycleNumber": "0"},
+):
+    camera = connectionTypes.PrerequisiteInput(
+        doc="Camera instrument",
+        name="camera",
+        storageClass="Camera",
+        dimensions=("instrument",),
+        isCalibration=True,
+    )
+    fgcm_visit_catalog = connectionTypes.Input(
+        doc="Catalog of visit information for fgcm.",
+        name="fgcmVisitCatalog",
+        storageClass="Catalog",
+        dimensions=("instrument",),
+    )
+    fgcm_fit_parameters_catalog = connectionTypes.Input(
+        doc="Catalog of fgcm fit parameters.",
+        name="fgcm_Cycle{cycleNumber}_FitParameters",
+        storageClass="Catalog",
+        dimensions=("instrument",),
+    )
+    flat = connectionTypes.PrerequisiteInput(
+        doc="Flat fields associated with illumination correction epoch.",
+        name="flat",
+        storageClass="ExposureF",
+        dimensions=["instrument", "detector", "physical_filter"],
+        isCalibration=True,
+        multiple=True,
+    )
+    illumination_corrections = connectionTypes.Output(
+        doc="Illumination corrections from fgcm fit.",
+        name="illuminationCorrection",
+        storageClass="ExposureF",
+        dimensions=["instrument", "detector", "physical_filter"],
+        isCalibration=True,
+        multiple=True,
+    )
+
+    def __init__(self, *, config=None):
+        super().__init__(config=config)
+
+        if str(int(config.connections.cycleNumber)) != config.connections.cycleNumber:
+            raise ValueError("cycleNumber must be of integer format")
+
+        if not config.use_flat_metadata:
+            del self.flat
+        else:
+            def lookup_flat(dataset_type, registry, data_id, collections):
+
+                time = Time(config.epoch_time, format=config.epoch_format)
+
+                return [
+                    registry.findDataset(
+                        dataset_type,
+                        data_id,
+                        physical_filter=physical_filter,
+                        timespan=Timespan(
+                            time - TimeDelta(1, format="sec"),
+                            time + TimeDelta(1, format="sec"),
+                        ),
+                        collections=collections,
+                    )
+                    for physical_filter in config.physical_filters
+                ]
```

Comment:
For `flat.metadata`, do:
```
return [
    registry.findDataset(
        dataset_type.makeComponentDatasetType(),
        data_id,
        ...
    ).makeComponentRef("metadata")
    for physical_filter in config.physical_filters
]
```

Reply:
I tried this again and now I recall the complaint:
```
lsst.daf.butler._exceptions_legacy.DatasetTypeError: Component dataset types are not supported in Registry methods; use DatasetRef or DatasetType methods to obtain components from parents instead.
```
And when I tried to build the ref myself something else got mad.  I think we'll just have to eat the inefficiency.

Reply:
I had a mistake in my snippet (now fixed; `Component` -> `Composite` in one place), if you're up for trying it again.